src/librustc/infer/error_reporting/mod.rs

```diff
@@ -607,22 +606,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         exp_found: Option<ty::error::ExpectedFound<Ty<'tcx>>>,
     ) {
         match cause.code {
-            ObligationCauseCode::MatchExpressionArmPattern { span, ty } => {
-                if ty.is_suggestable() {  // don't show type `_`
-                    err.span_label(span, format!("this match expression has type `{}`", ty));
-                }
-                if let Some(ty::error::ExpectedFound { found, .. }) = exp_found {
-                    if ty.is_box() && ty.boxed_ty() == found {
-                        if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
-                            err.span_suggestion(
-                                span,
-                                "consider dereferencing the boxed value",
-                                format!("*{}", snippet),
-                                Applicability::MachineApplicable,
-                            );
+            ObligationCauseCode::PatternGuard(box PatternGuardCause {
+                match_expr_discrim, other_range_expr
+            }) => {
+                if let Some((span, ty)) = match_expr_discrim {
+                    if ty.is_suggestable() {  // don't show type `_`
+                        err.span_label(
+                            span,
+                            format!("this match expression requires type `{}`", ty)
+                        );
+                    }
+                    if let Some(ty::error::ExpectedFound { found, .. }) = exp_found {
+                        if ty.is_box() && ty.boxed_ty() == found {
+                            if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                                err.span_suggestion(
+                                    span,
+                                    "consider dereferencing the boxed value",
+                                    format!("*{}", snippet),
+                                    Applicability::MachineApplicable,
+                                );
+                            }
                         }
                     }
                 }
+                if let Some((span, ty)) = other_range_expr {
+                    err.span_label(span, format!("other endpoint has type `{}`", ty));
```

Comment:
I feel like `"this endpoint has type` will lend itself to less confusion in the output.